### PR TITLE
Memory leaks + enhanced logging

### PR DESCRIPTION
**Pull request by marcstern**
Log audit lock name in case of problem.
Systematically log problems in update_rule_target_ex().
Fix some memory leaks in update_rule_target_ex().

### CODE REVIEW

**Comment by marcstern** [comment]
Replaces https://github.com/owasp-modsecurity/ModSecurity/pull/3156

File: apache2/re.c [hunk 1]
```diff
@@ -284,29 +282,22 @@ char *update_rule_target_ex(modsec_rec *msr, msre_ruleset *ruleset, msre_rule *r
...
                 name = param;
             }
 
-            if(apr_table_get(ruleset->engine->variables, name) == NULL)   {
-                if(target_list != NULL)
-                    free(target_list);
-                if(replace != NULL)
-                    free(replace);
-                if(msr) {
-                    msr_log(msr, 9, "Error to update target - [%s] is not valid target", name);
```
Comment (airween) [inline apache2/re.c:299]: Reviewed the logic and seems like this generated a duplicated error log entry, right?

I mean this line `msr_log()` and the next `return apr_psprintf()`.

If it's true, it's a fix that you removed or it caused by accident?

File: apache2/re.c [hunk 2]
```diff
@@ -372,28 +363,31 @@ char *update_rule_target_ex(modsec_rec *msr, msre_ruleset *ruleset, msre_rule *r
...
                     goto end;
                 }
-            } else {
+            }
+            else {
 
                 target = strdup(p);
-                if(target == NULL)
-                    return NULL;
+				if (target == NULL) {
```
Comment (airween) [inline apache2/re.c]: Seems like the indentation is wrong here. Could you correct it?

File: apache2/re.c [hunk 3]
```diff
@@ -408,30 +402,22 @@ char *update_rule_target_ex(modsec_rec *msr, msre_ruleset *ruleset, msre_rule *r
...
                     name = param;
                 }
 
-                if(apr_table_get(ruleset->engine->variables, name) == NULL)   {
-                    if(target_list != NULL)
-                        free(target_list);
-                    if(replace != NULL)
-                        free(replace);
-                    if(msr) {
-                        msr_log(msr, 9, "Error to update target - [%s] is not valid target", name);
```
Comment (airween) [inline apache2/re.c:424]: Same as above (in line 299. in old version)

**Review by airween — CHANGES_REQUESTED** [review]

File: apache2/re.c (hunk 1 above)
Comment (marcstern) [reply apache2/re.c:299]: You're right, it's a mistake.

File: apache2/re.c (hunk 2 above)
Comment (marcstern) [reply apache2/re.c]: I fixed it definitely in my editor

File: apache2/re.c (hunk 1 above)
Comment (airween) [reply apache2/re.c:299]: I don't know it's a mistake or not - this is why I asked (though it was a bit misunderstood).

`msr_log()` produces a **debug.log** entry.

But after calling the function chain, I'm not sure if it's also generated in that log, rather in **error.log**. The chain starts [here](https://github.com/marcstern/ModSecurity/blob/v2/pr/mem_leak_re/apache2/apache2_config.c#L3848) (in **your** branch).

How can we check this?

Comment (marcstern) [reply apache2/re.c:299]: We return `my_error_msg` which contains the error message (if any) that will be displayed in the error.log. That'll be done by httpd.
To check this, we could add a CI test with a buggy config.

File: apache2/re.c (hunk 3 above)
Comment (marcstern) [reply apache2/re.c:424]: Same answer as above

File: apache2/re.c (hunk 1 above)
Comment (airween) [reply apache2/re.c:299]: Could you add a buggy config to test this?

**Review by airween — APPROVED** [review]

**Pull request merged**